### PR TITLE
Initial implementation of extended profiles

### PR DESCRIPTION
**Pull request by erikjohnston**

### CODE REVIEW

File: synapse/federation/federation_server.py [hunk 1]
```diff
@@ -442,6 +442,17 @@ def on_get_missing_events(self, origin, room_id, earliest_events,
             "events": [ev.get_pdu_json(time_now) for ev in missing_events],
         })
 
+    def on_profile_request(self, user_id, persona, key):
+        if not self.hs.is_mine_id(user_id):
```
Comment (NegativeMjark) [inline synapse/federation/federation_server.py]: some Docstring might be nice.

File: synapse/storage/schema/delta/38/profile.py [hunk 2]
```diff
@@ -0,0 +1,104 @@
...
+
+def run_create(cur, database_engine, *args, **kwargs):
+    for statement in get_statements(CREATE_TABLE.splitlines()):
+        cur.execute(statement)
+
+
+def run_upgrade(cur, database_engine, config, *args, **kwargs):
+    if isinstance(database_engine, PostgresEngine):
+        cur.execute(UPDATE_DISPLAY_NAME, (config.server_name,))
+        cur.execute(UPDATE_AVATAR_URL, (config.server_name,))
```
Comment (NegativeMjark) [inline synapse/storage/schema/delta/38/profile.py]: If "UPDATE_.*" are postgres specific then maybe they should be called "POSTGRES_UPDATE_.*"?

File: synapse/storage/schema/delta/38/profile.py [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+logger = logging.getLogger(__name__)
+
+CREATE_TABLE = """
+CREATE TABLE profiles_extended (
+    stream_id BIGINT NOT NULL,
+    user_id TEXT NOT NULL,
+    persona TEXT NOT NULL,
+    key TEXT NOT NULL,
+    content TEXT NOT NULL
+);
```
Comment (NegativeMjark) [inline synapse/storage/schema/delta/38/profile.py]: Maybe comment on what the fields are?
Is the "content" JSON? Maybe comment on whether it is or not.

File: synapse/federation/federation_client.py [hunk 4]
```diff
@@ -878,6 +878,10 @@ def random_server_list():
 
         defer.returnValue(signed_events)
 
+    def get_profile(self, user_id, persona=None, key=None):
```
Comment (NegativeMjark) [inline synapse/federation/federation_client.py]: Docstring?

File: synapse/federation/transport/client.py [hunk 5]
```diff
@@ -396,3 +396,13 @@ def get_missing_events(self, destination, room_id, earliest_events,
         )
 
         defer.returnValue(content)
+
+    def get_profile(self, destination, user_id, persona=None, key=None):
```
Comment (NegativeMjark) [inline synapse/federation/transport/client.py]: Docstring

File: synapse/handlers/profile.py [hunk 6]
```diff
@@ -180,3 +180,29 @@ def _update_join_states(self, requester):
                     "Failed to update join event for room %s - %s",
                     j.room_id, str(e.message)
                 )
+
+    def get_full_profile_for_user(self, user_id):
```
Comment (NegativeMjark) [inline synapse/handlers/profile.py]: docstring?

File: synapse/handlers/profile.py [hunk 7]
```diff
@@ -180,3 +180,29 @@ def _update_join_states(self, requester):
...
+        if self.hs.is_mine_id(user_id):
+            return self.store.get_full_profile(user_id)
+        else:
+            return self.federation.get_profile(user_id)
+
+    def get_persona_profile_for_user(self, user_id, persona):
```
Comment (NegativeMjark) [inline synapse/handlers/profile.py]: docstring?

File: synapse/handlers/profile.py [hunk 8]
```diff
@@ -180,3 +180,29 @@ def _update_join_states(self, requester):
...
+        if self.hs.is_mine_id(user_id):
+            return self.store.get_persona_profile(user_id, persona)
+        else:
+            return self.federation.get_profile(user_id, persona)
+
+    def get_profile_key_for_user(self, user_id, persona, key):
```
Comment (NegativeMjark) [inline synapse/handlers/profile.py]: Docstring?

File: synapse/handlers/profile.py [hunk 9]
```diff
@@ -180,3 +180,29 @@ def _update_join_states(self, requester):
...
+        if self.hs.is_mine_id(user_id):
+            return self.store.get_profile_key(user_id, persona, key)
+        else:
+            return self.federation.get_profile(user_id, persona, key)
+
+    def update_profile_key(self, user_id, persona, key, content):
```
Comment (NegativeMjark) [inline synapse/handlers/profile.py]: Docstring?

File: synapse/storage/profile.py [hunk 10]
```diff
@@ -13,45 +13,159 @@
...
 class ProfileStore(SQLBaseStore):
     def create_profile(self, user_localpart):
-        return self._simple_insert(
-            table="profiles",
-            values={"user_id": user_localpart},
-            desc="create_profile",
+        return defer.succeed(None)
+
+    @defer.inlineCallbacks
+    def get_profile_displayname(self, user_id):
```
Comment (NegativeMjark) [inline synapse/storage/profile.py]: Docstring?

File: synapse/storage/profile.py [hunk 11]
```diff
@@ -13,45 +13,159 @@
...
         )
 
-    def set_profile_displayname(self, user_localpart, new_displayname):
-        return self._simple_update_one(
-            table="profiles",
-            keyvalues={"user_id": user_localpart},
-            updatevalues={"displayname": new_displayname},
-            desc="set_profile_displayname",
+    @defer.inlineCallbacks
+    def get_profile_avatar_url(self, user_id):
```
Comment (NegativeMjark) [inline synapse/storage/profile.py]: Docstring?

File: synapse/storage/profile.py [hunk 12]
```diff
@@ -13,45 +13,159 @@
...
+            try:
+                avatar_url = profile["rows"][0]["avatar_url"]
+            except (KeyError, IndexError):
+                avatar_url = None
+        else:
+            avatar_url = None
+
+        defer.returnValue(avatar_url)
+
+    def set_profile_avatar_url(self, user_id, new_avatar_url):
```
Comment (NegativeMjark) [inline synapse/storage/profile.py]: Docstring?

File: synapse/storage/profile.py [hunk 13]
```diff
@@ -13,45 +13,159 @@
...
         )
 
-    def set_profile_avatar_url(self, user_localpart, new_avatar_url):
-        return self._simple_update_one(
-            table="profiles",
-            keyvalues={"user_id": user_localpart},
-            updatevalues={"avatar_url": new_avatar_url},
-            desc="set_profile_avatar_url",
+    @defer.inlineCallbacks
+    def get_full_profile(self, user_id):
```
Comment (NegativeMjark) [inline synapse/storage/profile.py]: Docstring

File: synapse/storage/profile.py [hunk 14]
```diff
@@ -13,45 +13,159 @@
...
+        for row in rows:
+            content = ujson.loads(row["content"])
+            personas.setdefault(
+                row["persona"], {"rows": {}}
+            )["rows"][row["key"]] = content
+
+        defer.returnValue(profile)
+
+    @defer.inlineCallbacks
+    def get_persona_profile(self, user_id, persona):
```
Comment (NegativeMjark) [inline synapse/storage/profile.py]: Docstring

File: synapse/storage/profile.py [hunk 15]
```diff
@@ -13,45 +13,159 @@
...
+
+        persona = {"properties": {
+            row["key"]: ujson.loads(row["content"])
+            for row in rows
+        }}
+
+        defer.returnValue(persona)
+
+    @defer.inlineCallbacks
+    def get_profile_key(self, user_id, persona, key):
```
Comment (NegativeMjark) [inline synapse/storage/profile.py]: Docstring?

File: synapse/storage/profile.py [hunk 16]
```diff
@@ -13,45 +13,159 @@
...
+        )
+
+        if content_json:
+            content = ujson.loads(content_json)
+        else:
+            content = None
+
+        defer.returnValue(content)
+
+    def update_profile_key(self, user_id, persona, key, content):
```
Comment (NegativeMjark) [inline synapse/storage/profile.py]: Docstring?

**Comment by erikjohnston** [comment]
This isn't happening any time soon so closing.

**Pull request closed without merging**